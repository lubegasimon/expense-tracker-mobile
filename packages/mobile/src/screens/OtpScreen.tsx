import { useState } from "react";
import { View, StyleSheet } from "react-native";
import { useRoute } from "@react-navigation/native";
import { axiosInstance } from "@/src/api/axios";
import { resendCode, verifyCode } from "@/src/api/endpoints";
import Otp from "../components/Otp/Otp";
import OtpTitle from "../components/Otp/OtpTitle";
import ResendCodeSection from "../components/Otp/ResendCode";
import SubmitCodeButton from "../components/Otp/SubmitCode";
import ErrorBoundary from "../components/ErrorBoundary/ErrorBoundary";
import Loader from "../components/Loader/loader";
import Error from "../components/Error/Error";

function OtpScreen() {
  const [code, setCode] = useState("");
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState("");
  const [resentCode, setResentCode] = useState(false);

  interface EmailParam {
    email: string;
  }

  const route = useRoute();
  const { email } = route.params as EmailParam;

  function handleSubmit(code: string) {
    setLoading(true);
    axiosInstance
      .post(verifyCode, { code, email })
      .then(() => {
        setLoading(false);
        setError("");
      })
      .catch((error) => {
        setLoading(false);
        setError(error.response.data.error);
      });
  }

  function handleResendCode() {
    setLoading(true);
    axiosInstance
      .post(resendCode)
      .then(() => {
        setResentCode(true);
        setLoading(false);
        setError("");
      })
      .catch((error) => {
        setLoading(false);
        setError(error.response.data.error);
      });
  }

  if (loading) return <Loader />;
  else
    return (
      <ErrorBoundary>
        <View style={{ paddingTop: "10%" }}>
          <OtpTitle
            subTitle={
              resentCode
                ? "Code has been re-sent to your email"
                : "Code has been sent to your email"
            }
          />
          <Otp setCode={(code) => setCode(code)} />
          <Error error={error} style={errorStyle.error} />
          <ResendCodeSection onPress={handleResendCode} disabled={loading} />
          <SubmitCodeButton
            onPress={() => handleSubmit(code)}
            disabled={loading}
          />
        </View>
      </ErrorBoundary>
    );
}

const errorStyle = StyleSheet.create({
  error: {
    paddingLeft: "10%",
    paddingTop: "5%",
  },
});

export default OtpScreen;
