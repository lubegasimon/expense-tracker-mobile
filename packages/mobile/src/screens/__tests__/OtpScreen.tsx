import { render, screen } from "@testing-library/react-native";
import { NavigationContainer } from "@react-navigation/native";

import OtpScreen from "../OtpScreen";

jest.mock("@react-navigation/native", () => ({
  ...jest.requireActual("@react-navigation/native"),
  useRoute: () => ({
    params: { email: "johndoe@gmail.com" },
  }),
}));

describe("<OptScreen/>", () => {
  test("OtpScreen contains resend code and submit button", () => {
    const tree = render(
      <NavigationContainer>
        <OtpScreen />
      </NavigationContainer>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
    expect(screen.getByText("Resend code")).toBeTruthy();
    expect(screen.getByText("Submit")).toBeTruthy();
  });
});
