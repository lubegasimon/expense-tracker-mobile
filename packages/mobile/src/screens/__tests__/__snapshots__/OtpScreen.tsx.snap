// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OptScreen/> OtpScreen contains resend code and submit button 1`] = `
<View
  style={
    {
      "paddingTop": "10%",
    }
  }
>
  <Text
    style={
      {
        "fontSize": 30,
        "fontWeight": "bold",
        "paddingBottom": "3%",
        "paddingTop": "10%",
        "textAlign": "center",
      }
    }
  >
     Enter verification code 
  </Text>
  <Text
    style={
      {
        "fontSize": 15,
        "paddingBottom": "5%",
        "paddingLeft": "10%",
      }
    }
  >
     
    Code has been sent to your email
     
  </Text>
  <View>
    <View
      style={
        [
          {
            "flexDirection": "row",
            "justifyContent": "space-between",
            "width": "100%",
          },
          {
            "paddingLeft": "10%",
            "paddingRight": "10%",
          },
          undefined,
        ]
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "alignItems": "center",
              "borderColor": "#DFDFDE",
              "borderRadius": 12,
              "borderWidth": 1,
              "height": 60,
              "justifyContent": "center",
              "width": 44,
            },
            undefined,
            {
              "borderColor": "green",
            },
          ]
        }
        testID="otp-input"
      >
        <View
          collapsable={false}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              [
                {
                  "backgroundColor": "green",
                  "height": 30,
                  "width": 2,
                },
                {
                  "backgroundColor": "green",
                },
                undefined,
              ]
            }
            testID="otp-input-stick"
          />
        </View>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "alignItems": "center",
              "borderColor": "#DFDFDE",
              "borderRadius": 12,
              "borderWidth": 1,
              "height": 60,
              "justifyContent": "center",
              "width": 44,
            },
            undefined,
          ]
        }
        testID="otp-input"
      >
        <Text
          style={
            [
              {
                "fontSize": 28,
              },
              undefined,
              {},
            ]
          }
        />
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "alignItems": "center",
              "borderColor": "#DFDFDE",
              "borderRadius": 12,
              "borderWidth": 1,
              "height": 60,
              "justifyContent": "center",
              "width": 44,
            },
            undefined,
          ]
        }
        testID="otp-input"
      >
        <Text
          style={
            [
              {
                "fontSize": 28,
              },
              undefined,
              {},
            ]
          }
        />
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "alignItems": "center",
              "borderColor": "#DFDFDE",
              "borderRadius": 12,
              "borderWidth": 1,
              "height": 60,
              "justifyContent": "center",
              "width": 44,
            },
            undefined,
          ]
        }
        testID="otp-input"
      >
        <Text
          style={
            [
              {
                "fontSize": 28,
              },
              undefined,
              {},
            ]
          }
        />
      </View>
      <TextInput
        accessibilityLabel="One-Time Password"
        autoComplete="one-time-code"
        autoFocus={true}
        editable={true}
        inputMode="numeric"
        maxLength={4}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        secureTextEntry={false}
        style={
          [
            {
              "bottom": 0,
              "left": 0,
              "opacity": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            undefined,
          ]
        }
        testID="otp-input-hidden"
        textContentType="oneTimeCode"
        value=""
      />
    </View>
  </View>
  <View
    style={
      {
        "flexDirection": "row",
        "paddingLeft": "10%",
        "paddingTop": "3%",
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 18,
        }
      }
    >
      Didn't receive code?
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <Text
        style={
          {
            "color": "#34b4eb",
            "fontSize": 18,
            "paddingLeft": "2%",
          }
        }
      >
        Resend code
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "paddingLeft": "10%",
        "paddingRight": "10%",
        "paddingTop": "5%",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#34b4eb",
          "borderColor": "white",
          "borderRadius": 10,
          "padding": 10,
          "paddingRight": "10%",
          "width": "100%",
        }
      }
    >
      <Text
        style={
          {
            "color": "white",
            "fontSize": 25,
            "fontWeight": "bold",
            "textAlign": "center",
          }
        }
      >
        Submit
      </Text>
    </View>
  </View>
</View>
`;
